packages/flutter/lib/src/widgets/resolution_awareness.dart

```diff
@@ -0,0 +1,130 @@
+
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'media_query.dart';
+
+/// Asset resolver that knows how to find the asset most closely matching the
+/// current device pixel ratio, as given by [MediaQueryData].
+///
+/// Resolution aware assets are provided in groups. The group name is given
+/// as the asset key to the resolver, which then determines which specific
+/// asset in the group should be used.
+///
+/// For each group, a metadata asset should exist keyed by the group name +
+/// ".meta". This metadata asset is a textual list of individual assets for
+/// that group, separated by newlines. Individual assets should be keyed with a
+/// prefix of "Nx/" where N is the nominal device pixel ratio for that asset.
+/// N does not need to be an integer.
+///
+/// For example, suppose an application wants to use a logical icon named
+/// "heart". This icon has representations at 1.0, 1.5, and 2.0 pixel ratios. The
+/// asset bundle should then contain the following assets:
+///
+/// heart.meta
+/// 1.0x/heart.png
+/// 1.5x/heart.png
+/// 2.0x/heart.png
+///
+/// On a device with a 1.0 device pixel ratio, the image chosen would be
+/// 1.0x/heart.png; on a device with a 1.3 device pixel ratio, the image chosen
+/// would be 1.5x/heart.png.
+///
+/// Common prefixes will be removed, so the following would be a valid bundle
+/// layout for an asset group named "icons/heart":
+///
+/// icons/heart.meta
+/// icons/1.0x/heart.png
+/// icons/1.5x/heart.png
+/// icons/2.0x/heart.png
+class ResolutionAwareAssetResolver extends CandidateAssetResolver {
+  ResolutionAwareAssetResolver({ AssetBundle bundle, this.devicePixelRatio })
+    : super(bundle: bundle);
+
+  final double devicePixelRatio;
+
+  SplayTreeMap<double, String> _buildMapping(List<String> candidates) {
+    SplayTreeMap<double, String> result = new SplayTreeMap<double, String>();
+    RegExp exp = new RegExp(r"/?(\d+(\.\d*)?)x/");
+    for (String candidate in candidates) {
+      Match match = exp.firstMatch(candidate);
+      if (match != null && match.groupCount > 0) {
+        double resolution = double.parse(match.group(1));
+        result[resolution] = candidate;
+      }
+    }
+    return result;
+  }
+
+  String chooseCandidate(List<String> candidates) {
+    return new BreakpointQualifier(
+      candidates: _buildMapping(candidates),
+      value: devicePixelRatio
+    ).resolve();
+  }
+}
+
+/// Configuration for a ResolutionAwareAssetResolver.
+class ResolutionAwareResolverBuilder
+    extends AssetResolverBuilder<ResolutionAwareAssetResolver> {
+  const ResolutionAwareResolverBuilder({ this.devicePixelRatio, this.bundle });
+
+  final double devicePixelRatio;
+  final AssetBundle bundle;
+
+  ResolutionAwareAssetResolver buildResolver() {
+    return new ResolutionAwareAssetResolver(
+      bundle: bundle,
+      devicePixelRatio: devicePixelRatio
+    );
+  }
+
+  bool operator==(ResolutionAwareResolverBuilder config) {
```

Comment:
This doesn't quite fit the pattern for `operator==`.  You're supposed to omit the argument type and do an `is` check.  The idea is that `==` shouldn't ever throw.